### PR TITLE
add GaussianBlur augmentation

### PR DESCRIPTION
**Pull request by ZhiyuanChen**
### Description
Add augmentation of GaussianBlur

### Status
**Work in progress**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Review by edgarriba — CHANGES_REQUESTED** [review]
Thanks for the contribution. Could you provide minimal unit tests ? Don't forget to add the functionality to the documentation.

**Comment by ZhiyuanChen** [comment]
> Thanks for the contribution. Could you provide minimal unit tests ? Don't forget to add the functionality to the documentation.

Sure thing, but it could take a few days as I'm busy catching up ddl of CVPR.

File: kornia/augmentation/augmentation.py [hunk 1]
```diff
@@ -1148,3 +1148,48 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
...
+        >>> blur(input)
+    """
+    def __init__(self, kernel_size: Tuple[int, int],
+                 sigma: Tuple[float, float],
+                 border_type: str = 'reflect', 
+                 return_transform: bool = False,
+                 same_on_batch: bool = False,
+                 p: float = 0.5) -> None:
+        super(GaussianBlur, self).__init__(p=p, return_transform=return_transform, same_on_batch=same_on_batch, p_batch=p)
+        self.transform = K.filters.GaussianBlur2d(kernel_size, sigma, border_type)
```
Comment (shijianjian) [inline kornia/augmentation/augmentation.py:1189]: We are making all parameters trainable. Please refer to other implementations for doing so.

File: kornia/augmentation/augmentation.py [hunk 2]
```diff
@@ -1148,3 +1148,48 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
...
+    Note:
+        Input tensor must be float and normalized into [0, 1] for the best differentiability support.
+        Additionally, this function accepts another transformation tensor (:math:`(B, 3, 3)`), then the
+        applied transformation will be merged int to the input transformation tensor and returned.
+
+    Examples:
+        >>> rng = torch.manual_seed(0)
+        >>> input = torch.rand(1, 1, 5, 5)
+        >>> blur = GaussianBlur((3, 3), (0.1, 2.0), p=1.)
+        >>> blur(input)
```
Comment (shijianjian) [inline kornia/augmentation/augmentation.py:1180]: We are running doctest, so this should return the outcome.

File: kornia/augmentation/augmentation.py [hunk 3]
```diff
@@ -1148,3 +1148,48 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
...
+    """
+    def __init__(self, kernel_size: Tuple[int, int],
+                 sigma: Tuple[float, float],
+                 border_type: str = 'reflect', 
+                 return_transform: bool = False,
+                 same_on_batch: bool = False,
+                 p: float = 0.5) -> None:
+        super(GaussianBlur, self).__init__(p=p, return_transform=return_transform, same_on_batch=same_on_batch, p_batch=p)
```
Comment (shijianjian) [inline kornia/augmentation/augmentation.py:1188]:
```suggestion
        super(GaussianBlur, self).__init__(p=p, return_transform=return_transform, same_on_batch=same_on_batch, p_batch=1.)
```

File: kornia/augmentation/augmentation.py [hunk 4]
```diff
@@ -1148,3 +1148,48 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
 
     def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
         return F.apply_equalize(input, params)
+
```
Comment (shijianjian) [inline kornia/augmentation/augmentation.py:1151]: two blank lines to pass the static check.

**Review by shijianjian — CHANGES_REQUESTED** [review]

**Comment by edgarriba** [comment]
@ZhiyuanChen  We are releasing v0.5 in a week - would be great to include this feature.

**Comment by ZhiyuanChen** [comment]
> @ZhiyuanChen We are releasing v0.5 in a week - would be great to include this feature.

Copy, sorry it was quite busy for the past two weeks, I'll try to finish them this weekend.

File: kornia/augmentation/augmentation.py (hunk 1 above)
Comment (ZhiyuanChen) [reply kornia/augmentation/augmentation.py:1189]: Can I ask what parameters should be trainable here? sigma?

Comment (shijianjian) [reply kornia/augmentation/augmentation.py]: Theoretically, yes. If the user input sigma as ```nn.Parameter(1.)```, it will be trainable (the user may use ```xxx.backward()``` to update it if they want). Otherwise, it will be static in our current design.

Comment (ZhiyuanChen) [reply kornia/augmentation/augmentation.py]: Thank you for your clarification. Can I ask if there are any other parameters should be trainable?

Comment (shijianjian) [reply kornia/augmentation/augmentation.py]: Currently, only sigma shall be trainable since it is a continuous variable.

For kernel size and borderType, they are not going to be trainable for now since they are discrete variables. At some point, we did think about putting somewhat gradient estimation for the discrete variables (for augmentations), but I think it is not going to happen very soon. 

If you are interested in this, please feel free to have a PR or an issue to discuss more about it :)

Comment (edgarriba) [reply kornia/augmentation/augmentation.py]: to make sigma to be trainable, some changes are needed here: https://github.com/kornia/kornia/blob/master/kornia/filters/kernels.py#L20-L25
but this should be done in a separated PR since might imply other changes I guess with typing. I would stick this PR to sigma with floats and leave for later the support to trainable tensors. We should consider later too to add too some sort of random gaussian blur/noise/etc.

Comment (ZhiyuanChen) [reply kornia/augmentation/augmentation.py]: Ok, I'll resolve this for now, the reset changes to do is some more tests and the docs?

**Comment by edgarriba** [comment]
@ZhiyuanChen could you verify that your rebased correctly from the upstream project ? I suggest to do as follows: git pull --rebase upstream master - solve conflicts and you might have to force push.

**Comment by ZhiyuanChen** [comment]
> @ZhiyuanChen could you verify that your rebased correctly from the upstream project ? I suggest to do as follows: git pull --rebase upstream master - solve conflicts and you might have to force push.

Done

**Review by edgarriba — APPROVED** [review]
LGTM. You can add better unit tests later

**Comment by edgarriba** [comment]
@shijianjian resolve conflicts and merge

**Review by shijianjian — APPROVED** [review]
